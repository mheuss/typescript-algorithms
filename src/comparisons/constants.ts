export enum Comparisons {
  LessThan = -1,
  Equal,
  GreaterThan,
}

export type ComparisonFunction<T> = (a: T, b: T) => Comparisons;
