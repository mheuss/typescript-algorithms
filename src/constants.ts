export enum ErrorCodes {
  COMPARISON_FUNCTION_REQUIRED = 1,
  LIST_IS_BROKEN_INTERNALLY,
  OPERATION_BEYOND_BOUNDS,
  FAILED_TO_SET,
  COMPARISON_ERROR,
  INVALID_TREE_DATA_FILE,
}
