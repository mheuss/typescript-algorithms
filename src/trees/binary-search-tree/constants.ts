export enum Sides {
  Left,
  Right,
}
