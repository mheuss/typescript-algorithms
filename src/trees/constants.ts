export interface ITreeObject<T> {
  children: Array<ITreeObject<T>>;
  value: T;
}

export enum Traversal {
  Breadth_First,
  DepthFirstInOrder,
  DepthFirstPreOrder,
  DepthFirstPostOrder,
}
