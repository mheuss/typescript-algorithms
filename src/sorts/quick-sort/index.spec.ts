import { ErrorCodes } from '../../constants';
import { QuickSort } from './index';

describe('Merge Sort tests', () => {
  it('Should properly sort a list of numbers', () => {
    expect(
      QuickSort<number>([76, 23, 97, 12, 54, 5, 0, -1, 2])
    ).toEqual([-1, 0, 2, 5, 12, 23, 54, 76, 97]);
  });

  it('Should just return if given a signle number in an array', () => {
    expect(
      QuickSort<number>([1])
    ).toEqual([1]);
  });

  it('Should properly sort a list of strings', () => {
    expect(
      QuickSort<string>([
        'pineapple',
        'artichoke',
        'celery',
        'eggplant',
        'banana',
        'Kiwi',
      ])
    ).toMatchSnapshot();
  });

  it('Should properly sort objects', () => {
    interface INames {
      lastName: string;
      firstName: string;
    }

    const arr = [
      { lastName: 'Heuss', firstName: 'Mike' },
      { lastName: 'Smith', firstName: 'John' },
      { lastName: 'Arnold', firstName: 'Peggy' },
      { lastName: 'Saben', firstName: 'Matt' },
      { lastName: 'LaCroix', firstName: 'Mark' },
    ];
    const results = QuickSort<INames>(arr, (a: INames, b: INames) => {
      return a.lastName.localeCompare(b.lastName);
    });
    expect(results.length).toEqual(arr.length);
    expect(results).toMatchSnapshot();
  });

  it('Should return an empty array if given an empty array', () => {
    const results = QuickSort<number>([]);
    expect(results).toEqual([]);
  });

  it('Should throw if array is of type it cannot sort, and no sort function was passed in', () => {
    const arr = [
      { lastName: 'Heuss', firstName: 'Mike' },
      { lastName: 'Smith', firstName: 'John' },
      { lastName: 'Arnold', firstName: 'Peggy' },
      { lastName: 'Saben', firstName: 'Matt' },
      { lastName: 'LaCroix', firstName: 'Mark' },
    ];
    try {
      QuickSort(arr);
      fail('Should have thrown');
    } catch (e) {
      expect(e.code).toEqual(ErrorCodes.COMPARISON_FUNCTION_REQUIRED);
    }
  });
});
