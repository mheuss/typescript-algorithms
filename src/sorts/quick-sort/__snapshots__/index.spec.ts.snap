// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merge Sort tests Should properly sort a list of strings 1`] = `
Array [
  "artichoke",
  "banana",
  "celery",
  "eggplant",
  "Kiwi",
  "pineapple",
]
`;

exports[`Merge Sort tests Should properly sort objects 1`] = `
Array [
  Object {
    "firstName": "Peggy",
    "lastName": "Arnold",
  },
  Object {
    "firstName": "Mike",
    "lastName": "Heuss",
  },
  Object {
    "firstName": "Mark",
    "lastName": "LaCroix",
  },
  Object {
    "firstName": "Matt",
    "lastName": "Saben",
  },
  Object {
    "firstName": "John",
    "lastName": "Smith",
  },
]
`;
